src/algo.cpp

```diff
@@ -0,0 +1,230 @@
+#include "algo.h"
+#include "game_battler.h"
+#include "game_actor.h"
+#include "game_enemy.h"
+#include "attribute.h"
+#include "player.h"
+#include <lcf/rpg/skill.h>
+
+#include <algorithm>
+
+namespace Algo {
+
+bool IsRowAdjusted(const Game_Actor& actor, lcf::rpg::System::BattleCondition cond, bool offense) {
+	return (cond == lcf::rpg::System::BattleCondition_surround
+			|| (actor.GetBattleRow() == (1 - offense)
+				&& (cond == lcf::rpg::System::BattleCondition_none || cond == lcf::rpg::System::BattleCondition_initiative))
+			|| (actor.GetBattleRow() == offense
+				&& (cond == lcf::rpg::System::BattleCondition_back))
+		   );
+}
+
+int CalcNormalAttackToHit(const Game_Battler &source, const Game_Battler &target, lcf::rpg::System::BattleCondition cond) {
+	auto to_hit = source.GetHitChance();
+
+	// If target has rm2k3 state which grants 100% dodge.
+	if (target.EvadesAllPhysicalAttacks()) {
+		return 0;
+	}
+
+	// If target has Restriction "do_nothing", the attack always hits
+	if (!target.CanAct()) {
+		return 100;
+	}
+
+	// Modify hit chance for each state the source has
+	to_hit = (to_hit * source.GetHitChanceModifierFromStates()) / 100;
+
+	// Stop here if attacker ignores evasion.
+	if (source.GetType() == Game_Battler::Type_Ally
+		&& static_cast<const Game_Actor&>(source).AttackIgnoresEvasion()) {
+		return to_hit;
+	}
+
+	// AGI adjustment.
+	to_hit = 100 - (100 - to_hit) * (1.0f + (float(target.GetAgi()) / float(source.GetAgi()) - 1.0f) / 2.0f) ;
+
+	// If target has physical dodge evasion:
+	if (target.GetType() == Game_Battler::Type_Ally
+			&& static_cast<const Game_Actor&>(target).HasPhysicalEvasionUp()) {
+		to_hit -= 25;
+	}
+
+	// Defender row adjustment
+	if (Player::IsRPG2k3() && target.GetType() == Game_Battler::Type_Ally) {
+		if (IsRowAdjusted(static_cast<const Game_Actor&>(target), cond, false)) {
+			to_hit -= 25;
+		} else if(source.GetType() == Game_Battler::Type_Ally
+				&& target.GetType() == Game_Battler::Type_Enemy
+				&& cond == lcf::rpg::System::BattleCondition_back) {
+			// FIXME: RPG_RT always adjusts damage down for back attack, regardless of row - when to handle and not handle this?
+			to_hit -= 25;
+		}
+	}
+
+	return to_hit;
+}
+
+
+int CalcSkillToHit(const Game_Battler& source, const Game_Battler& target, const lcf::rpg::Skill& skill) {
+	auto to_hit = skill.hit;
+
+	if (skill.failure_message != 3
```

Comment:
> Use Algo library and port normal attack calculations to it

Could you flag this as a RPG2k3 bug? (the skill message can't be modified)

Reply:
I think this doesn't apply anymore. Saw this in a later commit commented.

Reply:
It was added later, but it should go above here to make it easier to read.